ambry-api/src/main/java/com.github.ambry/router/ReadableStreamChannel.java

```diff
@@ -45,4 +46,33 @@
    * @return the {@link Future} that will eventually contain the result of the operation.
    */
   public Future<Long> readInto(AsyncWritableChannel asyncWritableChannel, Callback<Long> callback);
+
+  /**
+   * Set the digest algorithm to use on the data that is being streamed from the channel. Once the channel is emptied,
+   * the digest can be obtained via {@link #getDigest()}.
+   * <p/>
+   * This function is ideally called before {@link #readInto(AsyncWritableChannel, Callback)}. After a call to
+   * {@link #readInto(AsyncWritableChannel, Callback)}, some content may have been discarded and getting a digest may
+   * longer be possible. The safety of doing otherwise depends on the implementation.
+   * @param digestAlgorithm the digest algorithm to use.
+   * @throws NoSuchAlgorithmException if the {@code digestAlgorithm} does not exist or is not supported.
+   * @throws IllegalStateException if {@link #readInto(AsyncWritableChannel, Callback)} has already been called.
+   */
+  public void setDigestAlgorithm(String digestAlgorithm)
+      throws NoSuchAlgorithmException;
+
+  /**
+   * Gets the digest as specified by the digest algorithm set through {@link #setDigestAlgorithm(String)}. If none was
+   * set, returns {@code null}.
+   * <p/>
+   * This function is ideally called after the channel is emptied completely. Otherwise, the complete digest may not
+   * have been calculated yet. The safety of doing otherwise depends on the implementation.
+   * <p/>
+   * "Emptying a channel" refers to awaiting on the future or getting the callback after a
+   * {@link #readInto(AsyncWritableChannel, Callback)} call.
+   * @return the digest as specified by the digest algorithm set through {@link #setDigestAlgorithm(String)}. If none
```

Comment:
my thinking was that the algorithm cannot have "computed" - either is fine with me. Which do you think is clearer?